File: divtest/lib/resultsStore.js
/**
 * Persist backtest results as one JSON file per instrument/timeframe/month.
 * Layout: data/{instrument}/{timeframe}/{YYYY-MM}.json
 */
'use strict';

const fs = require('fs');
const path = require('path');
const { getConfig } = require('./config');
const { analyzeTrades } = require('./strategyEngine');

function safeName(s) {
  return String(s || 'UNKNOWN')
    .toUpperCase()
    .replace(/[^A-Z0-9._-]+/g, '_');
}

function safeTimeframe(tf) {
  return String(tf || '15min')
    .toLowerCase()
    .replace(/[^a-z0-9._-]+/g, '_');
}

function monthFile(instrument, timeframe, yyyyMm) {
  const root = getConfig().dataDir;
  return path.join(root, safeName(instrument), safeTimeframe(timeframe), `${yyyyMm}.json`);
}

function ensureDir(filePath) {
  fs.mkdirSync(path.dirname(filePath), { recursive: true });
}

function readMonth(instrument, timeframe, yyyyMm) {
  const fp = monthFile(instrument, timeframe, yyyyMm);
  if (!fs.existsSync(fp)) return null;
  try {
    return JSON.parse(fs.readFileSync(fp, 'utf8'));
  } catch {
    return null;
  }
}

function monthExists(instrument, timeframe, yyyyMm) {
  return fs.existsSync(monthFile(instrument, timeframe, yyyyMm));
}

function writeMonth(instrument, timeframe, yyyyMm, payload) {
  const fp = monthFile(instrument, timeframe, yyyyMm);
  ensureDir(fp);
  const doc = {
    instrument: safeName(instrument),
    timeframe,
    month: yyyyMm,
    updated_at: new Date().toISOString(),
    source: payload.source || null,
    instrument_key: payload.instrument_key || null,
    candles: payload.candles || 0,
    notes: payload.notes || [],
    trades: payload.trades || [],
    meta: payload.meta || {},
  };
  fs.writeFileSync(fp, JSON.stringify(doc, null, 2));
  return doc;
}

/**
 * Append trades into a month file (idempotent by entry+exit+side key when merging runs).
 */
function appendMonthTrades(instrument, timeframe, yyyyMm, trades, extra = {}) {
  const existing = readMonth(instrument, timeframe, yyyyMm) || {
    trades: [],
    notes: [],
    candles: 0,
  };
  const keyOf = (t) =>
    `${t.side}|${t.entry_datetime}|${t.exit_datetime}|${t.entry_price}|${t.exit_price}`;
  const seen = new Set((existing.trades || []).map(keyOf));
  const merged = [...(existing.trades || [])];
  for (const t of trades || []) {
    const k = keyOf(t);
    if (!seen.has(k)) {
      seen.add(k);
      merged.push(t);
    }
  }
  return writeMonth(instrument, timeframe, yyyyMm, {
    ...existing,
    ...extra,
    trades: merged,
    notes: [...new Set([...(existing.notes || []), ...(extra.notes || [])])],
  });
}

function listInstruments() {
  const root = getConfig().dataDir;
  if (!fs.existsSync(root)) return [];
  return fs
    .readdirSync(root, { withFileTypes: true })
    .filter((d) => d.isDirectory())
    .map((d) => d.name)
    .sort();
}

function listMonths(instrument, timeframe) {
  const dir = path.join(getConfig().dataDir, safeName(instrument), safeTimeframe(timeframe));
  if (!fs.existsSync(dir)) return [];
  return fs
    .readdirSync(dir)
    .filter((f) => /^\d{4}-\d{2}\.json$/.test(f))
    .map((f) => f.replace(/\.json$/, ''))
    .sort()
    .reverse();
}

/**
 * Load all stored trades with optional filters.
 */
function loadAllTrades({ instrument = null, timeframe = null } = {}) {
  const root = getConfig().dataDir;
  if (!fs.existsSync(root)) return [];
  const instruments = instrument ? [safeName(instrument)] : listInstruments();
  const tfs = timeframe && timeframe !== 'All' ? [timeframe] : ['10min', '15min', '1hr'];
  const trades = [];
  const files = [];

  for (const inst of instruments) {
    for (const tf of tfs) {
      const dir = path.join(root, inst, safeTimeframe(tf));
      if (!fs.existsSync(dir)) continue;
      for (const f of fs.readdirSync(dir)) {
        if (!f.endsWith('.json')) continue;
        const fp = path.join(dir, f);
        files.push(fp);
        try {
          const doc = JSON.parse(fs.readFileSync(fp, 'utf8'));
          for (const t of doc.trades || []) {
            trades.push({ ...t, instrument: t.instrument || inst, timeframe: t.timeframe || tf, _month: doc.month });
          }
        } catch {
          /* skip corrupt */
        }
      }
    }
  }

  trades.sort((a, b) => String(a.entry_datetime).localeCompare(String(b.entry_datetime)));
  return { trades, files: files.length };
}

function getResultsSummary(filters = {}) {
  const { trades } = loadAllTrades(filters);
  const analysis = analyzeTrades(trades);
  const byTimeframe = {};
  for (const tf of ['10min', '15min', '1hr']) {
    const subset = trades.filter((t) => t.timeframe === tf);
    byTimeframe[tf] = analyzeTrades(subset);
  }
  return {
    filters,
    instruments: listInstruments(),
    trade_count: trades.length,
    trades,
    analysis,
    by_timeframe: byTimeframe,
  };
}

module.exports = {
  safeName,
  monthFile,
  readMonth,
  monthExists,
  writeMonth,
  appendMonthTrades,
  listInstruments,
  listMonths,
  loadAllTrades,
  getResultsSummary,
  safeTimeframe,
};

File: backend/services/divtest/results_store.py
"""JSON month files: data/divtest/{instrument}/{timeframe}/{YYYY-MM}.json"""
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any, Dict, List, Optional

from backend.services.divtest.config import DATA_DIR, TIMEFRAMES
from backend.services.divtest.strategy import analyze_trades


def _safe(name: str) -> str:
    return re.sub(r"[^A-Z0-9._-]+", "_", str(name or "UNKNOWN").upper())


def _tf(tf: str) -> str:
    return re.sub(r"[^a-z0-9._-]+", "_", str(tf or "15min").lower())


def month_path(instrument: str, timeframe: str, yyyy_mm: str) -> Path:
    return DATA_DIR / _safe(instrument) / _tf(timeframe) / f"{yyyy_mm}.json"


def month_exists(instrument: str, timeframe: str, yyyy_mm: str) -> bool:
    return month_path(instrument, timeframe, yyyy_mm).exists()


def read_month(instrument: str, timeframe: str, yyyy_mm: str) -> Optional[Dict[str, Any]]:
    p = month_path(instrument, timeframe, yyyy_mm)
    if not p.exists():
        return None
    try:
        return json.loads(p.read_text(encoding="utf-8"))
    except Exception:
        return None


def write_month(instrument: str, timeframe: str, yyyy_mm: str, payload: Dict[str, Any]) -> Dict[str, Any]:
    p = month_path(instrument, timeframe, yyyy_mm)
    p.parent.mkdir(parents=True, exist_ok=True)
    from datetime import datetime, timezone

    doc = {
        "instrument": _safe(instrument),
        "timeframe": _tf(timeframe),
        "month": yyyy_mm,
        "updated_at": datetime.now(timezone.utc).isoformat(),
        "source": payload.get("source"),
        "instrument_key": payload.get("instrument_key"),
        "candles": payload.get("candles") or 0,
        "notes": payload.get("notes") or [],
        "trades": payload.get("trades") or [],
        "meta": payload.get("meta") or {},
    }
    p.write_text(json.dumps(doc, indent=2), encoding="utf-8")
    return doc


def append_month_trades(
    instrument: str, timeframe: str, yyyy_mm: str, trades: List[Dict[str, Any]], extra: Optional[Dict[str, Any]] = None
) -> Dict[str, Any]:
    existing = read_month(instrument, timeframe, yyyy_mm) or {"trades": [], "notes": [], "candles": 0}
    extra = extra or {}

    def key_of(t: Dict[str, Any]) -> str:
        return f"{t.get('side')}|{t.get('entry_datetime')}|{t.get('exit_datetime')}|{t.get('entry_price')}|{t.get('exit_price')}"

    seen = {key_of(t) for t in (existing.get("trades") or [])}
    merged = list(existing.get("trades") or [])
    for t in trades or []:
        k = key_of(t)
        if k not in seen:
            seen.add(k)
            merged.append(t)
    notes = list({*(existing.get("notes") or []), *(extra.get("notes") or [])})
    return write_month(
        instrument,
        timeframe,
        yyyy_mm,
        {
            **existing,
            **extra,
            "trades": merged,
            "notes": notes,
        },
    )


def list_instruments() -> List[str]:
    if not DATA_DIR.exists():
        return []
    return sorted([p.name for p in DATA_DIR.iterdir() if p.is_dir()])


def load_all_trades(instrument: Optional[str] = None, timeframe: Optional[str] = None) -> Dict[str, Any]:
    instruments = [_safe(instrument)] if instrument and instrument != "All" else list_instruments()
    tfs = [_tf(timeframe)] if timeframe and timeframe != "All" else [t["id"] for t in TIMEFRAMES]
    trades: List[Dict[str, Any]] = []
    files = 0
    for inst in instruments:
        for tf in tfs:
            d = DATA_DIR / inst / tf
            if not d.exists():
                continue
            for f in d.glob("*.json"):
                files += 1
                try:
                    doc = json.loads(f.read_text(encoding="utf-8"))
                except Exception:
                    continue
                for t in doc.get("trades") or []:
                    row = dict(t)
                    row.setdefault("instrument", inst)
                    row.setdefault("timeframe", tf)
                    row["_month"] = doc.get("month")
                    trades.append(row)
    trades.sort(key=lambda t: str(t.get("entry_datetime") or ""))
    return {"trades": trades, "files": files}


def get_results_summary(instrument: Optional[str] = None, timeframe: Optional[str] = None) -> Dict[str, Any]:
    loaded = load_all_trades(instrument=instrument, timeframe=timeframe)
    trades = loaded["trades"]
    analysis = analyze_trades(trades)
    by_tf = {}
    for tf in [t["id"] for t in TIMEFRAMES]:
        subset = [t for t in trades if t.get("timeframe") == tf]
        by_tf[tf] = analyze_trades(subset)
    return {
        "filters": {"instrument": instrument or "All", "timeframe": timeframe or "All"},
        "instruments": list_instruments(),
        "trade_count": len(trades),
        "trades": trades,
        "analysis": analysis,
        "by_timeframe": by_tf,
    }
